### ISSUE TITLE
nxagent segfaults on session resume

### ISSUE THREAD

**Issue by HolySephi** [ArcticaProject/nx-libs#118]
Our environment:
- Ubuntu 14.04
- X2Go from PPA/X2Go
- Arctica from `deb http://packages.arctica-project.org/ubuntu-nightly trusty main`

nxagent segfaults reproducable after a few resumes with the same message every time.

Steps to reproduce:
1. `/usr/bin/nxagent :1`
2. get the pid of the nxagent process
3. `while :; do kill -SIGHUP <pid>; sleep 10; done`
     -> the segfault appears faster with sleep 3 - but even with 10 seconds (to give the process enough time to finish every action of resuming) the bug is appearing
4. Wait for the segmentation fault:

nxagent output:

```
Session: Resuming session at 'Mon Apr 25 13:00:27 2016'.
nxagentSetScreenSaverTime: ScreenSaverTime was [600000], ScreenSaverInterval was [600000].
nxagentSetScreenSaverTime: ScreenSaverTime now is [600000], ScreenSaverInterval now is [600000].
nxagentOpenScreen: Called for screen index [0].
nxagentOpenScreen: Root window at offset (0,0) size (2880,810).
nxagentOpenScreen: Default window at offset (2,139) size (2880,810) border size 0.
nxagentOpenScreen: Span between root window and default window is (0,0).
nxagentOpenScreen: Default window in window mode has offset (0,0) and size (2880,810).
nxagentOpenScreen: Fullscreen is OFF.
nxagentOpenScreen: Desktop resize mode is ON.
nxagentOpenScreen: Resize desktop at startup is OFF.
nxagentOpenScreen: Going to create new default window.
nxagentOpenScreen: Created new default window with id [8388686].
nxagentOpenScreen: Setting WM_CLASS and WM_NAME for window withid [8388686].
nxagentOpenScreen: Found WM, delete window atom [240].
Info: Not using local device configuration changes.
nxagentSaveScreen: Called for screen at [0x17aed50] with parameter [3].
nxagentSaveScreen: SCREEN_SAVER_ON is [0] SCREEN_SAVER_OFF is [1] SCREEN_SAVER_FORCER is [2] SCREEN_SAVER_CYCLE is [3].
SessionPath not defined
The XKEYBOARD keymap compiler (xkbcomp) reports:
> Error:            Cannot open "compiled/server-1.xkm" to write keyboard description
>                   Exiting
Session: Session resumed at 'Mon Apr 25 13:00:29 2016'.
Segmentation fault (core dumped)
```

gdb output:

```
Session: Session resumed at 'Mon Apr 25 12:38:42 2016'.

Program received signal SIGHUP, Hangup.

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff78dbdcf in XPeekEvent (dpy=dpy@entry=0xc38f60, event=event@entry=0x7fffffffd970) at PeekEvent.c:49
49      PeekEvent.c: No such file or directory.
(gdb)
```

[See attached file for full gdb log](https://github.com/ArcticaProject/nx-libs/files/234381/gdb_log_nxagent_segfault.txt)

**Comment by uli42** [comment]
Please try to reproduce this with -extension RANDR

**Comment by HolySephi** [comment]
As stated in #117 adding "-extension RANDR" changes nothing in log output or behaviour

**Comment by uli42** [comment]
But xdpyinfo does not show the extensions anymore, right?

**Comment by HolySephi** [comment]
xdpyinfo neither states RANDR nor XINERAMA; correct

**Comment by uli42** [comment]
I can reproduce this with a debbugging version (and sleep .1 instead of sleep 10) and get this:

> nxagentDispatchEvents: Going to handle new event type [15].
> nxagentDispatchEvents: Going to handle new VisibilityNotify event.
> nxagentDispatchEvents: Visibility notify state is [1] with previous [0].
> nxagentDispatchEvents: Going to handle new event type [22].
> nxagentDispatchEvents: Going to handle new ConfigureNotify event.
> nxagentWaitEvents called.
> _X11TransSocketBytesReadable(0xce6e30,5,0x7fffffffb1e0)
> _XFlushInt: Entering flush with [0] bytes to write.
> _X11TransSocketRead(5,0x7fffffffb200,32)
> _XWaitForReadable: Calling select().
> _XWaitForReadable: WAIT! Waiting for the display to become readable.
> _XSelect: Called with [6][0x7fffffffb0c0][(nil)][(nil)][0x7ffff7bc3480].
> 
> Program received signal SIGHUP, Hangup.
> _XWaitForReadable: Out of select with result [-1] and errno [4].
> _XIOError: Resetting the display buffer.
> _XIOError: Resetting the display flags.
> _XIOError: Resetting the display buffer.
> _XIOError: Resetting the display flags.
> _XReadEvents: Returning with I/O error detected.
> 
> Program received signal SIGSEGV, Segmentation fault.
> 0x00007ffff78d485f in XPeekEvent (dpy=dpy@entry=0xc72300, event=event@entry=0x7fffffffd2b0) at > PeekEvent.c:49
> 49        *event = (dpy->head)->event;
> (gdb) bt
> #0  0x00007ffff78d485f in XPeekEvent (dpy=dpy@entry=0xc72300, event=event@entry=0x7fffffffd2b0) at PeekEvent.c:49
> #1  0x000000000048f647 in nxagentWaitEvents (tm=0x7fffffffd290, dpy=0xc72300) at Events.c:4560
> #2  nxagentHandleConfigureNotify (X=X@entry=0x7fffffffd4e0) at Events.c:3390
> #3  0x0000000000491358 in nxagentDispatchEvents (predicate=predicate@entry=0x0) at Events.c:1948
> #4  0x000000000049c31a in nxagentWakeupHandler (data=<optimized out>, count=1, mask=0xb49ca0 <LastSelectMask>) at Handlers.c:631
> #5  0x000000000045bd3b in WakeupHandler (result=result@entry=1, pReadmask=0xb49ca0 <LastSelectMask>) at dixutils.c:472
> #6  0x000000000046dfd5 in WaitForSomething (pClientsReady=pClientsReady@entry=0x7fffffffd900) at WaitFor.c:389
> #7  0x000000000042f70b in Dispatch () at NXdispatch.c:619
> #8  0x000000000040e0d5 in main (argc=2, argv=0x7fffffffdec8, envp=<optimized out>) at main.c:441
> (gdb) list
> 44        register XEvent *event;
> 45    {
> 46        LockDisplay(dpy);
> 47        if (dpy->head == NULL)
> 48            _XReadEvents(dpy);
> 49        *event = (dpy->head)->event;
> 50        UnlockDisplay(dpy);
> 51        return 1;
> 52    }
> 53  
> (gdb) print dpy
> $1 = (Display *) 0xc72300
> (gdb) print dpy->head
> $2 = (struct _XSQEvent *) 0x0`

So, XReadEvents() does not set dpy->head, for whatever reason.

**Comment by uli42** [comment]
XReadEvents() has NX modifications, that - under some circumstances - return where the original code did not return. If that happens in a state where no event has been read, this could explain the problem.

My log above shows that this branch was executed just before segfault:

> #ifdef NX_TRANS_SOCKET
>             if (dpy->flags & XlibDisplayIOError)
>             {
> #if defined(NX_TRANS_SOCKET) && defined(NX_TRANS_DEBUG)
>                 fprintf(stderr, "_XReadEvents: Returning with I/O error detected.\n");
> #endif
>                 return;
>             }
> #endif

Must be analyzed further. Help appreciated ;-)

**Comment by uli42** [comment]
I can confirm that commit 2d776b14fddc5ec70c97aa82672f3a7c9caef6a3 does not show that segfault, but the next one (c6482d24fff982d5e11636167c719766f8207d48) has this (and a similar segfault, too). I'll try to find the reason the next days.

**Comment by sunweaver** [comment]
@uli42: Thanks for investigating this and taking your time for fixing it. :-)

**Comment by uli42** [comment]
I have not really come further although I investigated quite some time....

And I am no longer sure that the problem is not in 2d776b14fddc5ec70c97aa82672f3a7c9caef6a3. It might just be more difficult to trigger.

The culprit is that XPeekEvent() calls XReadEvents(), which should block until at least one event is in the queue. But the nx code has added some "return" statements where the original code has none. So it returns and XPeekEvent(), which called XReadEvents(), expects an event to be there, but there's none. It _seems_ that there's is an IOError flag set somewhere before that is triggering this in  XReadEvents().

We could add a check in XPeekEvent() to prevent the crash at least but I am not sure how to go on in the case of no event being around. And upstream does not check it either.

I suspect multithreading but I am not really sure if there are really multiple threads running, mut check that. I found some posting from 2006 stating that XLib is not threadsafe although it claims so. The suggestion was to use different displays for different threads, whatever that means in detail. Also I found that there's no call of XInitThreads() but adding it did not change anything.

So I am a bit unsure about the next steps. If anybody could take a look at lib/X11/XlibInt.c , especially _XReadEvents() I'd appreciate that.

**Comment by uli42** [comment]
Today I have tested some more and came across the following output, which leads to the conclusion that gettign the SIGHUP _during_ Xselect() leads to all this: 

> _XWaitForReadable: Calling select().
> _XWaitForReadable: WAIT! Waiting for the display to become readable.
> _XSelect: Called with [7][0x7fffffffb0c0][(nil)][(nil)][0x7ffff7bc3480].
> 
> Program received signal SIGHUP, Hangup.
> _XWaitForReadable: Out of select with result [-1] and errno [4].
> _XIOError: Resetting the display buffer.
> _XIOError: Resetting the display flags.
> _XIOError: Resetting the display buffer.
> _XIOError: Resetting the display flags.
> _XReadEvents: Returning with I/O error detected.
> 
> Program received signal SIGSEGV, Segmentation fault.
> 0x00007ffff78d961f in XPeekEvent (dpy=dpy@entry=0xbbc400, event=event@entry=0x7fffffffd2b0) at > PeekEvent.c:49
> 49        *event = (dpy->head)->event;
> (gdb) bt
> #0  0x00007ffff78d961f in XPeekEvent (dpy=dpy@entry=0xbbc400, event=event@entry=0x7fffffffd2b0) at PeekEvent.c:49
> #1  0x000000000048ead7 in nxagentWaitEvents (tm=0x7fffffffd290, dpy=0xbbc400) at Events.c:4560
> #2  nxagentHandleConfigureNotify (X=X@entry=0x7fffffffd4e0) at Events.c:3390
> #3  0x00000000004907e8 in nxagentDispatchEvents (predicate=predicate@entry=0x0) at Events.c:1948
> #4  0x000000000049b7f7 in nxagentWakeupHandler (data=<optimized out>, count=-1, mask=0xb48c60 <LastSelectMask>) at Handlers.c:653
> #5  0x000000000045b10b in WakeupHandler (result=result@entry=-1, pReadmask=0xb48c60 <LastSelectMask>) at dixutils.c:472
> #6  0x000000000046d3a5 in WaitForSomething (pClientsReady=pClientsReady@entry=0x7fffffffd900) at WaitFor.c:389
> #7  0x000000000042f79b in Dispatch () at NXdispatch.c:620
> #8  0x000000000040e125 in main (argc=2, argv=0x7fffffffdec8, envp=<optimized out>) at main.c:441

**Comment by uli42** [comment]
Thinking a bit about this problem it came to my mind that this way of using nxagent is not really a use case. SIGHUP is used for disconnects and reconnects when the nx transport is used. And looking at the code it is a different codepath that handles the signal then (see XSelect() in XlibInt.c). It is questionable if using nxagent in the described way has to be supported at all. I mean, you can use Xnest or Xephyr if you do not use the nx transport anyway.

So we could make the SIGHUP handler ignoring the signal in that case. I have implemented that and it prevents the crashes while disabling the ability to send an nxagent without nx transport to sleep. (BTW: the same code path is used for `-nopersistent`).

Another signal could still cause this EINTR during select() problem but I have not managed to trigger it (yet?).

@HolySephi: what is your use case for this?

**Comment by sunweaver** [comment]
Hi Uli,

On  So 01 Mai 2016 22:12:43 CEST, Ulrich Sibiller wrote:

> Thinking a bit about this problem it came to my mind that this way  
> of using nxagent is not really a use case. SIGHUP is used for  
> disconnects and reconnects when the nx transport is used. And  
> looking at the code it is a different codepath that handles the  
> signal then (see XSelect() in XlibInt.c). It is questionable if  
> using nxagent in the described way has to be supported at all. I  
> mean, you can use Xnest or Xephyr if you do not use the nx transport  
> anyway.

It is not a use case now, but it may become one.

Actually, I use nxagent instead of Xephyr and Xnest everywhere,  
because it performs so much better than the others (esp. regarding  
resizing, regarding Xinerama).

I actually also suspend sessions via SIGHUP sometimes.

Furthermore, we plan to use freerdp-shadow-x11 on top of nxagent (at  
least it is an idea) and x11vnc. Once we do that, we will not use  
nxcomp as transport for all kinds of sessions. The nxagent process  
will become more an Xserver like Xnest/Xephyr that supports being  
connected to via nxcomp.

> So we could make the SIGHUP handler ignoring the signal in that  
> case. I have implemented that and it prevents the crashes while  
> disabling the ability to send an nxagent without nx transport to  
> sleep. (BTW: you can get the same behaviour by using `-nopersistent`

Can't we protect the tricky code passages from being caught by  
unexpected SIGHUP signals?

> Another signal could still cause this EINTR during select() problem  
> but I have not managed to trigger it (yet?).

## Mike

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Key ID 0x25771B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

freeBusy:
https://mail.das-netzwerkteam.de/mailxchange/kronolith/fb.php?u=m.gabriel%40das-netzwerkteam.de

**Comment by HolySephi** [comment]
My use-case is that x2go with arctica is sometimes segfaulting on resume. The kill -SIGHUP procedure was a fast reproducing advice from sunweaver and to be able to gdb the segfault correctly.

I don't know if it is this specific segfault but there is sometimes a segfault on resuming sessions with x2go.

**Comment by uli42** [comment]
@sunweaver: Well, this would have to be done in XlibInt.c:_XSelect(), I guess. The first part of _XSelect() is taking care of the nx transport, the second part is dealing with the non-nx transport. So the second part should be adapted accordingly. But how? If you get EINTR you could re-enter a loop that repeat the select(). However, when EINTR is returned you should also check what signal you got and eventually leave the loop. As signals are handled in nxagent (and not the lib): how can we access the needed variables in a clean way from the library?

If you simply ignore the EINTR and return 0 (same like when the timeout is reached) in that case the nxagent's IOError handler also being triggered...

Please have a look at it, I might be thinking  too complicated again...;-)

@HolySephi: by disabling the nx transport you are using a different code path so this is probably not the same segfault.

**Comment by HolySephi** [comment]
I'm sorry but i don't understand - how did i disable the nx transport? Is manually running nxagent so different from spawning nxagent from a x2goclient? Is suspending and resuming sessions via x2go different from sending SIGHUP to the nxagent process?

**Comment by uli42** [comment]
By running nxagent locally without specifying an nx display (nx/nx...) the nx transport is disabled/not activated. You can force usage of the transport by specifying `-forcenx`, IIRC.

**Comment by HolySephi** [comment]
so then starting nxagent like:
`nxagent -forcenx :1`
?

Because the same segfault still occurs while sending SIGHUP

**Comment by sunweaver** [comment]
Hi Florian,

On  Mo 02 Mai 2016 09:42:18 CEST, Florian Wicke wrote:

> I'm sorry but i don't understand - how did i disable the nx  
> transport? Is manually running nxagent so different from spawning  
> nxagent from a x2goclient? Is suspending and resuming sessions via  
> x2go different from sending SIGHUP to the nxagent process?

A DDX like Xnest or Xephyr only speak the X11 proto and directly  
communicate with an underlying X(.org )server.

A DDX like nxagent has two options: (a) communicate directly with an  
underlying X(.org )server or (b) communicate with an underlying  
X(.org) server through nxagent<->nxcomp(<->nxproxy).

Greets,
Mike

## 

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Key ID 0x25771B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

freeBusy:
https://mail.das-netzwerkteam.de/mailxchange/kronolith/fb.php?u=m.gabriel%40das-netzwerkteam.de

**Comment by uli42** [comment]
Ok, hmm, interesting, must check that again.

**Comment by sunweaver** [comment]
Hi Uli,

On  Mo 02 Mai 2016 07:57:38 CEST, Ulrich Sibiller wrote:

> Please have a look at it, I might be thinking  too complicated again...;-)

I may not have grasped the underlying problem fully, but I would start  
with looking at
nxagentSighupHandler() in hw/nxagent/Display.c.

It seems we can manipulate when SIGHUP is supposed to suspend a  
session and when not.

Also, the SIGHUP signal triggering a session suspension is not limited  
to nxagent using the nxcomp transport. It also works with nxagent  
connecting to X.org server directly.

But the overall question is... Why can't this issue be triggered with  
nx-libs 3.5.x whereas it can be triggered (obviously) with 3.6.x.

## Mike

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Key ID 0x25771B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

freeBusy:
https://mail.das-netzwerkteam.de/mailxchange/kronolith/fb.php?u=m.gabriel%40das-netzwerkteam.de

**Comment by sunweaver** [comment]
Hi Uli,

On  Mo 02 Mai 2016 07:57:38 CEST, Ulrich Sibiller wrote:

> @sunweaver: Well, this would have to be done in  
> XlibInt.c:_XSelect(), I guess. The first part of _XSelect() is  
> taking care of the nx transport, the second part is dealing with the  
> non-nx transport. So the second part should be adapted accordingly.  
> But how? If you get EINTR you could re-enter a loop that repeat the  
> select(). However, when EINTR is returned you should also check what  
> signal you got and eventually leave the loop. As signals are handled  
> in nxagent (and not the lib): how can we access the needed variables  
> in a clean way from the library?

we may not be able to access any variables from nxagent within the lib.

However, maybe we can hold the incoming SIGHUP signal for some time  
and wait until an ongoing _XSelect operation has been processed fully  
and suspending is safe.

## Mike

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Key ID 0x25771B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

freeBusy:
https://mail.das-netzwerkteam.de/mailxchange/kronolith/fb.php?u=m.gabriel%40das-netzwerkteam.de

**Comment by HolySephi** [comment]
Just to add as much information as possible to this case - i ran
`nxagent :1`
with the current 3.5.0 "stable" of nxagent and SIGHUPed the hell out of it for about half an hour now every 5s without running into the segfault i get after 2-10 SIGHUPs with arctica nxagent

even changing the interval to .2 did not segfault the nxagent in any way

**Comment by uli42** [comment]
@Mike: Are you familiar with signal handling? My googling the last days revealed a lot of details to be aware of. + differences between OSes.

@HolySephi: so did I, without "success". I even tried 0.1s.. However, disabling all the xrandr/xinerama code in 3.6.x did not help either... still crashes without that code.

**Comment by sunweaver** [comment]
Can you please build nx-libs from source from this branch and test if the error also occurs using that branch?
https://github.com/sunweaver/nxlibs, Branch: pr/hw-nxagent-drop-duplicate-Xserver-code_disable-Xserver-code-and-include-in-nxagent

```
  git clone https://github.com/sunweaver/nx-libs.git
  cd nx-libs
  git fetch origin pr/hw-nxagent-drop-duplicate-Xserver-code_disable-Xserver-code-and-include-in-nxagent
  git checkout pr/hw-nxagent-drop-duplicate-Xserver-code_disable-Xserver-code-and-include-in-nxagent
  debuild -uc -us
  cd ..
  sudo dpkg -i *.deb
```

Greets,
Mike

**Comment by sunweaver** [comment]
Ah... forgot to mention... I have tested the above and had 400+ resume/suspends using HolySephi's online-torturer.

**Comment by uli42** [comment]
Interesting. Let's say that we really cannot reproduce the error with that changes. What does that mean for us? Is there some code that has not been linked in before because it existed twice?

Will try that this evening (hopefully...)

**Comment by HolySephi** [comment]
I built the packages as requested with your extra patches but the segmentation fault still occurs.

**Comment by uli42** [comment]
Same here, segfault almost immediately

Update: and going back to e4763fb4879ce8782e50c784ed9e7d8c5a1b7322 the torture produces

> Session: Session resumed at 'Mon May  2 23:26:27 2016'.
> 
> Program received signal SIGHUP, Hangup.
> Xlib: sequence lost (0x10000 > 0x119) in reply type 0x4!
> nxagentHandleReparentNotify: WARNING! XGetWindowAttributes failed.
> 
> Program received signal SIGSEGV, Segmentation fault.
> 0x000000000048fe65 in nxagentDispatchEvents (predicate=predicate@entry=0x0) at Events.c:1251
> 1251            x.u.u.detail = inputInfo.pointer -> button -> map[nxagentReversePointerMap[X.xbutton.button - 1]];

**Comment by sunweaver** [comment]
Some more test cases:
- my latest code from https://github.com/sunweaver/nx-libs/tree/pr/hw-nxagent-drop-duplicate-Xserver-code_disable-Xserver-code-and-include-in-nxagent
- Debian jessie client (X.org from jessie)
- Debian jessie nxagent (built from above sources)
- Debian sid nxagent (built from above sources)
- Ubuntu trusty nxagnet (built from above sources and launched inside chroot)
- Ubuntu trusty nxagent running in Ubuntu trusty Xephyr

Window managers used on client-side: i3, marco from the MATE Desktop, openbox.

I cannot reproduce the segfaults you report above.

**Comment by sunweaver** [comment]
... and all systems are 64bit.

**Comment by uli42** [comment]
Now it is getting interesting! What's the version of the X server you using on trusty? Can you please retry with Unity?

**Comment by uli42** [comment]
I tried to compile on SuSE 13.1 and failed:

> In file included from ../../dix/dixfonts.c:59:0,
>                  from NXdixfonts.c:77:
> /usr/include/X11/fonts/fontmisc.h:92:13: error: conflicting types for ‘CopyISOLatin1Lowered’
>  extern void CopyISOLatin1Lowered(
>              ^
> In file included from ../../include/dixstruct.h:29:0,
>                  from NXdixfonts.c:71:
> ../../include/dix.h:384:13: note: previous declaration of ‘CopyISOLatin1Lowered’ was here
>  extern void CopyISOLatin1Lowered(
>              ^

That's the drawback of using the system libraries...

**Comment by sunweaver** [comment]
On  Di 03 Mai 2016 17:49:30 CEST, Ulrich Sibiller wrote:

> I tried to compile onn SuSE 13.1 and failed:
> In file included from ../../dix/dixfonts.c:59:0,
>                  from NXdixfonts.c:77:
> /usr/include/X11/fonts/fontmisc.h:92:13: error: conflicting types  
> for ‘CopyISOLatin1Lowered’
>  extern void CopyISOLatin1Lowered(
>              ^
> In file included from ../../include/dixstruct.h:29:0,
>                  from NXdixfonts.c:71:
> ../../include/dix.h:384:13: note: previous declaration of  
> ‘CopyISOLatin1Lowered’ was here
>  extern void CopyISOLatin1Lowered(
>              ^

@uli42: please re-fetch the branch [1] anew so it includes this commit:
https://github.com/sunweaver/nx-libs/commit/e3e451629b00ab589f8720d498089dba0447cf8f

The previous approach (that included <X11/fonts/fontmisc.h> caused  
FTBFS on systems using libXfont << 1.5).

The new approach is much better.

Mike

[1]  

## https://github.com/sunweaver/nx-libs/commits/pr/hw-nxagent-drop-duplicate-Xserver-code_disable-Xserver-code-and-include-in-nxagent

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Key ID 0x25771B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

freeBusy:
https://mail.das-netzwerkteam.de/mailxchange/kronolith/fb.php?u=m.gabriel%40das-netzwerkteam.de

**Comment by sunweaver** [comment]
I actually force-pushed to the same branch again just now.

**Comment by sunweaver** [comment]
Tested with a very minimal Unity desktop shell in a chroot running in Xepyhr. All from Ubuntu 14.04 64bit.

I can reproduce the segfaults then.

**Comment by uli42** [comment]
Ok, what now? So Unity must do something the other WMs do not.

**Comment by uli42** [comment]
I was able to compile the recent version on Suse 13.1 and did not get the segfault in XPeekEvent but (again) this:

> Program received signal SIGHUP, Hangup.
> Xlib: sequence lost (0x10000 > 0xcd) in reply type 0x4!
> nxagentHandleReparentNotify: WARNING! XGetWindowAttributes failed.
> 
> Program received signal SIGSEGV, Segmentation fault.
> 0x000000000049092d in nxagentDispatchEvents (predicate=predicate@entry=0x0) at > Events.c:1255
> 1255              x.u.u.detail = inputInfo.pointer -> button -> map[nxagentReversePointerMap[X.xbutton.button - 1]];
> (gdb) bt
> #0  0x000000000049092d in nxagentDispatchEvents (predicate=predicate@entry=0x0) > at Events.c:1255
> #1  0x000000000049b30f in nxagentWakeupHandler (data=<optimized out>, count=1, mask=0xb33680 <LastSelectMask>) at Handlers.c:631
> #2  0x00000000004590cb in WakeupHandler (result=result@entry=1, pReadmask=0xb33680 <LastSelectMask>) at dixutils.c:472
> #3  0x000000000046ea15 in WaitForSomething (pClientsReady=pClientsReady@entry=0x7fffffffd8b0) at WaitFor.c:389
> #4  0x000000000043054a in Dispatch () at NXdispatch.c:357
> #5  0x000000000040f1e9 in main (argc=2, argv=0x7fffffffde68, envp=<optimized out>) at main.c:441

and when running with `-sync` I get this (again):

> Program received signal SIGSEGV, Segmentation fault.
> 0x00007ffff76e185f in XPeekEvent (dpy=dpy@entry=0xe1c100, event=event@entry=0x7fffffffd060) at PeekEvent.c:49
> 49              *event = (dpy->head)->event;
> (gdb) bt
> #0  0x00007ffff76e185f in XPeekEvent (dpy=dpy@entry=0xe1c100, event=event@entry=0x7fffffffd060) at PeekEvent.c:49
> #1  0x000000000048f50e in nxagentWaitEvents (tm=0x7fffffffd040, dpy=0xe1c100) at Events.c:4525
> #2  nxagentHandleConfigureNotify (X=X@entry=0x7fffffffd260) at Events.c:3356
> #3  0x0000000000490fba in nxagentDispatchEvents (predicate=predicate@entry=0x0) at Events.c:1942
> #4  0x000000000049b93c in nxagentDispatchHandler (client=client@entry=0x0, in=in@entry=0, out=out@entry=0) at Handlers.c:1288
> #5  0x00000000004a74b4 in nxagentSynchronizeRegion (pDrawable=0xdc0ac0, pRegion=<optimized out>, breakMask=7, owner=0x0) at Drawable.c:929
> #6  0x00000000004a91d5 in nxagentSynchronizeDrawableData (pDrawable=0xdc0ac0, breakMask=7, owner=0x0) at Drawable.c:356
> #7  0x00000000004a9def in nxagentSynchronizeDrawable (owner=<optimized out>, breakMask=<optimized out>, wait=<optimized out>, pDrawable=0x0) at Drawable.c:169
> #8  nxagentSynchronizeDrawablePredicate (p0=0xdc0ac0, x1=8, p2=0x1, p2@entry=0x7fffffffd640) at Drawable.c:1288
> #9  0x0000000000447815 in FindClientResourcesByType (client=0xb4bf90, type=13, func=func@entry=0x4a9cd0 <nxagentSynchronizeDrawablePredicate>, cdata=cdata@entry=0x7fffffffd640) at NXresource.c:433
> #10 0x00000000004a55b1 in nxagentSynchronizationLoop (mask=mask@entry=7) at Drawable.c:1419
> #11 0x000000000049af7a in nxagentBlockHandler (data=<optimized out>, timeout=0x7fffffffd740, mask=<optimized out>) at Handlers.c:341
> #12 0x0000000000458fda in BlockHandler (pTimeout=pTimeout@entry=0x7fffffffd740, pReadmask=0xb33680 <LastSelectMask>) at dixutils.c:442
> #13 0x000000000046e91e in WaitForSomething (pClientsReady=pClientsReady@entry=0x7fffffffd8a0) at WaitFor.c:257
> #14 0x000000000043054a in Dispatch () at NXdispatch.c:357

So it does not seem to be limited to Unity, this was with KDE in a nested nx session.

**Comment by uli42** [comment]
With this patch I did not see the crashes. However, this is ugly...

```
diff --git a/nx-X11/lib/X11/PeekEvent.c b/nx-X11/lib/X11/PeekEvent.c
index b89c6c9..38b6347 100644
--- a/nx-X11/lib/X11/PeekEvent.c
+++ b/nx-X11/lib/X11/PeekEvent.c
@@ -46,7 +46,8 @@ XPeekEvent (dpy, event)
        LockDisplay(dpy);
        if (dpy->head == NULL)
            _XReadEvents(dpy);
-       *event = (dpy->head)->event;
+        if (dpy->head != NULL)
+         *event = (dpy->head)->event;
        UnlockDisplay(dpy);
        return 1;
}
```

**Comment by uli42** [comment]
I think the source of the problem is Unity placing the nxagent window at a slightly different position for every resume. So a ConfigureNotify event gets triggered. Handling that event is probably unexpected at some parts of the code.

**Comment by uli42** [comment]
Configuring a fixed position for nxagent (using https://wiki.ubuntuusers.de/Compiz_CCSM/ ) greatly prolonged the time until I got a segfault. Next I disabled window animation and fading in ccsm and could not produce a single segfault. But interestingly reverting all these changes did not bring back the segfaults! I still cannot get any now! Can anyone confirm this?

**Comment by HolySephi** [comment]
But i don't use unity or compiz - in my tests the nxagent was displayed via X11 forward(from ubuntu14.04) on my Xming(Windows 10/MobaXterm)

**Comment by uli42** [comment]
Oh, even more places to look for the problem ;-(

However, are you able to test the ugly patch I posted yesterday? It might
lead to other problems but at least it will prevent the segfault.
Am 08.05.2016 02:19 schrieb "Florian Wicke" notifications@github.com:

**Comment by HolySephi** [comment]
i'll be able to test the patch on tuesday. I don't mind if you don't test nxagent and Xming - our setup in special is x2goagent running inside a plain Xorg-process (like the x2gothinclient setup) and it is segfaulting there too.

Maybe your approach was the wrong way? Maybe it is segfaulting in "normal" or most setups but not in the special env of sunweaver?

**Comment by uli42** [comment]
Yes, of course that's possible... However, I could also not reproduce this without Unity.

Vom Smartphone gesendet.

----- Ursprüngliche Nachricht -----
Von: "Florian Wicke" notifications@github.com
Gesendet: ‎08.‎05.‎2016 16:56
An: "ArcticaProject/nx-libs" nx-libs@noreply.github.com
Cc: "Ulrich Sibiller" uli42@gmx.de; "Mention" mention@noreply.github.com
Betreff: Re: [ArcticaProject/nx-libs] nxagent segfaults on session resume(#118)

i'll be able to test the patch on tuesday. I don't mind if you don't test nxagent and Xming - our setup in special is x2goagent running inside a plain Xorg-process (like the x2gothinclient setup) and it is segfaulting there too.
Maybe your approach was the wrong way? Maybe it is segfaulting in "normal" or most setups but not in the special env of sunweaver?

**Comment by HolySephi** [comment]
With the patch from above i'm no longer able to run into the segfault but the screen stays gray sometimes (i think this would be time where the segfault would trigger) instead of getting the "black" background of the empty X session.

**Comment by HolySephi** [comment]
I'm currently testing this code:

```
~# diff -c3 a/nx-X11/lib/X11/PeekEvent.c nx-libs/nx-X11/lib/X11/PeekEvent.c
*** a/nx-X11/lib/X11/PeekEvent.c        2016-05-11 13:58:53.057080686 +0000
--- b/nx-X11/lib/X11/PeekEvent.c        2016-05-11 13:25:14.912208689 +0000
***************
*** 44,50 ****
        register XEvent *event;
  {
        LockDisplay(dpy);
!       if (dpy->head == NULL)
            _XReadEvents(dpy);
        *event = (dpy->head)->event;
        UnlockDisplay(dpy);
--- 44,50 ----
        register XEvent *event;
  {
        LockDisplay(dpy);
!       while (dpy->head == NULL)
            _XReadEvents(dpy);
        *event = (dpy->head)->event;
        UnlockDisplay(dpy);
```

and it looks promising so far

**Comment by HolySephi** [comment]
with my code the session always gets resumed (sometimes it takes a few extra seconds) but if you suspend the session before the session was able to get the _XReadEvents(dpy); then the process will hang forever

**Comment by HolySephi** [comment]
the screen stays gray between these two log entries from nxagent:

```
Session: Session resumed at 'Wed May 11 13:44:44 2016'.
  <here is the pause for 1-20 seconds where the display is only gray until the display wents "black" for the empty x session>
nxagentChangeScreenConfig: WindowTable[0] is 0x17dfb30
```

**Comment by sunweaver** [comment]
Hi,

On  Mi 11 Mai 2016 15:34:28 CEST, Florian Wicke wrote:

> I'm currently testing this code:
> 
> ```
>         LockDisplay(dpy);
>         while (dpy->head == NULL)
>             _XReadEvents(dpy);
>        *event = (dpy->head)->event;
>         UnlockDisplay(dpy);
> ```
> 
> and it looks promising so far

Shouldn't this go into the nxagent code where XPeekEvent is called?

We avoid hacking libX11...

## Mike

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Key ID 0x25771B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

freeBusy:
https://mail.das-netzwerkteam.de/mailxchange/kronolith/fb.php?u=m.gabriel%40das-netzwerkteam.de

**Comment by uli42** [comment]
It should not be done at all, that is why I was calling that 'ugly'. The main problem is that XReadEvents() returns without having read any event. It should be blocking but it is not thanks to code added by nx.

Besides: Busy looping is not nice either...

Vom Smartphone gesendet.

----- Ursprüngliche Nachricht -----
Von: "Mike Gabriel" notifications@github.com
Gesendet: ‎11.‎05.‎2016 20:11
An: "ArcticaProject/nx-libs" nx-libs@noreply.github.com
Cc: "Ulrich Sibiller" uli42@gmx.de; "Mention" mention@noreply.github.com
Betreff: Re: [ArcticaProject/nx-libs] nxagent segfaults on session resume(#118)

Hi,

On Mi 11 Mai 2016 15:34:28 CEST, Florian Wicke wrote:

> I'm currently testing this code:
> 
> ```
> LockDisplay(dpy);
> while (dpy->head == NULL)
> _XReadEvents(dpy);
> *event = (dpy->head)->event;
> UnlockDisplay(dpy);
> ```
> 
> and it looks promising so far

Shouldn't this go into the nxagent code where XPeekEvent is called?

We avoid hacking libX11...

## Mike

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Key ID 0x25771B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

freeBusy:
https://mail.das-netzwerkteam.de/mailxchange/kronolith/fb.php?u=m.gabriel%40das-netzwerkteam.de

**Comment by HolySephi** [comment]
Of course this was just a quickndirty hack - i thought it might be helpful to test this to get the information if dpy stays NULL on a segfaulted resume or will get filled later on. It seems that _XReadEvents(dpy); needs more time sometimes and should block/loop for itself then.

The unchanged code now just reads it once if its NULL and then magically thinks it will work without further checks.

**Comment by sunweaver** [comment]
It seems that 3.5.0.x is also affected: http://bugs.x2go.org/cgi-bin/bugreport.cgi?bug=908

**Comment by HolySephi** [comment]
I encountered segfaults in 3.5.0.x very very rarely and was unable to reproduce the segfault like we did in 3.6.0.x - since i limited the maximum session time to one day  i never encountered any hanging or segfaulting sessions with 3.5.0.x

**Comment by uli42** [comment]
I have again looked into that. I think the nx modification of XReadEvents() is buggy.

```
/* _XReadEvents - Flush the output queue,                                                                                                              
 * then read as many events as possible (but at least 1) and enqueue them                                                                              
 */
```

But there are two code paths where the nx version simply returns without ensuring that there's at least one event read and enqueued. I am unsure what to do about that... Should we insert a fake event in that case?

**Comment by uli42** [comment]
I think I've found a workaround. Please try this:

```
diff --git a/nx-X11/programs/Xserver/hw/nxagent/Events.c b/nx-X11/programs/Xserver/hw/nxagent/Events.c
index c04481c..4233546 100644
--- a/nx-X11/programs/Xserver/hw/nxagent/Events.c
+++ b/nx-X11/programs/Xserver/hw/nxagent/Events.c
@@ -4557,7 +4557,8 @@ int nxagentWaitEvents(Display *dpy, struct timeval *tm)
   }
   else
   {
-    XPeekEvent(dpy, &ev);
+    if (Xpending(dpy))
+       XPeekEvent(dpy, &ev);
   }

   /*
```

**Comment by HolySephi** [comment]
Unfortunately this patch renders the code un-build-able:

`/root/nx-libs/nx-X11/programs/Xserver/hw/nxagent/Events.c:4560: undefined reference to 'Xpending'`

**Comment by uli42** [comment]
Typo, try a capital P

Vom Smartphone gesendet.

----- Ursprüngliche Nachricht -----
Von: "Florian Wicke" notifications@github.com
Gesendet: ‎22.‎06.‎2016 17:36
An: "ArcticaProject/nx-libs" nx-libs@noreply.github.com
Cc: "Ulrich Sibiller" uli42@gmx.de; "Mention" mention@noreply.github.com
Betreff: Re: [ArcticaProject/nx-libs] nxagent segfaults on session resume(#118)

Unfortunately this patch renders the code un-build-able:
/root/nx-libs/nx-X11/programs/Xserver/hw/nxagent/Events.c:4560: undefined reference toXpending'
`

**Comment by sunweaver** [comment]
@uli42: Just heard back from @HolySephi on IRC. Patch seems to work. Please file a PR for this fix. (Or shall I?).

**Comment by uli42** [comment]
Nice to hear, at least it helps. But I used the evening to rethink this again and came to the conclusion that this patch is wrong. It is wrong because it changes the semantics. XPeekEvent() is supposed to block, but if we only run it when there are events waiting the code will never block. So in the normal case, where no signals or IOErrors are around, it should really block. With my patch it will never block...

**Comment by sunweaver** [comment]
Hi Ulrich,

On  Do 23 Jun 2016 09:40:13 CEST, Ulrich Sibiller wrote:

> Nice to hear, at least it helps. But I used the evening to rethink  
> this again and came to the conclusion that this patch is wrong. It  
> is wrong because it changes the semantics. XPeekEvent() is supposed  
> to block, but if we only run it when there are events waiting the  
> code will never block. So in the normal case, where no signals or  
> IOErrors are around, it should really block. With my patch it will  
> never block...

I think that the non-blocking hacks are actually what NoMachine came  
up with that makes nx-libs superior to Xnest.

I understand that this all is not required when using XCB instead of Xlib.

What happens with the nxagent performance when we introduce blocking  
code in XPeekEvent?

Actually, NoMachine also added a function to Xlib that also does  
something around events: XCheckIfEventNoFlush. This function call  
reads from the event queue without flushing it. Also very "peculiar"  
and non-Xlib like.

Mike

## 

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22  0782 9AF4 6B30 2577 1B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

**Comment by HolySephi** [comment]
Yeah as sunweaver said i'm unable to run into any segfaults with this change. Neither 'sleep .1' nor 'sleep 10' are able to produce segmentation faults over the testing period of half an hour each.

**Comment by sunweaver** [comment]
On  Do 23 Jun 2016 09:54:36 CEST, Florian Wicke wrote:

> Yeah as sunweaver said i'm unable to run into any segfaults with  
> this change. Neither 'sleep .1' nor 'sleep 10' are able to produce  
> segmentation faults over the testing period of half an hour each.

How about the overall performance of the running session? How about  
the process load of nxagent comparing the version without uli42's  
workaround and with?

## Mike

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22  0782 9AF4 6B30 2577 1B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

**Comment by uli42** [comment]
Once again: with this patch we are handling a special case and for that special case it really helps.

However, it changes the normal case. The normal case at this part of the code (nxagentWaitEvents(), which is described as "Blocks until an event becomes available.") is a call of XPeekEvent() and expecting it to block until an event occurs. When we call XPending() before XPeekEvent() and only call XPeekEvent() if there are events there will never be any blocking in nxagentWaitEvents()). So we need to decide at every nxagentWaitEvents() call if blocking behaviour is required or not. But how to decide?

Second, the nx version of _XReadEvents() is simply broken because it breaks the semantics of the function. Any call to that function is likely to cause problems if it depends on the original behaviour. Who can tell if in a future version further calls to _XReadEvents() appear in the upstream Xlib code?

So the real fix would be to find out why nx has added these return to _XReadEvents() in the first place and if there are other ways to get the desired effect without breaking the semantics.

Maybe instead of simply returning we could insert fake events.

@sunweaver:
"I think that the non-blocking hacks are actually what NoMachine came up with that makes nx-libs superior to Xnest." 

Superior regarding what?

"I understand that this all is not required when using XCB instead of Xlib."

Maybe, but this is out of focus for now.

"What happens with the nxagent performance when we introduce blocking code in XPeekEvent?"

XPeekEvent already is blocking.

"Actually, NoMachine also added a function to Xlib that also does something around events: XCheckIfEventNoFlush. This function call reads from the event queue without flushing it. Also very "peculiar" and non-Xlib like."

Yeah, and there are lots of other functions in this area:
http://www.sbin.org/doc/Xlib/chapt_08.html -> "Routines that Get Events"

**Comment by sunweaver** [comment]
Hi Uli,

On  Do 23 Jun 2016 10:12:14 CEST, Ulrich Sibiller wrote:

> Once again: with this patch we are handling a special case and for  
> that special case it really helps.

Ok.

> However, it changes the normal case. The normal case at this part of  
> the code (nxagentWaitEvents(), which is described as "Blocks until  
> an event becomes available.") is a call of XPeekEvent() and  
> expecting it to block until an event occurs. When we call XPending()  
> before XPeekEvent() and only call XPeekEvent() if there are events  
> there will never be any blocking in nxagentWaitEvents()). So we need  
> to decide at every nxagentWaitEvents() call if blocking behaviour is  
> required or not. But how to decide?

So would it make sense to introduce a looping around XPending()  
directly in nxagentWaitEvents()? Instead of leaving the blocking to  
NX_X11, we could move that into nxagent code directly.

> Second, the nx version of _XReadEvents() is simply broken because it  
> breaks the semantics of the function. Any call to that function is  
> likely to cause problems if it depends on the original behaviour.  
> Who can tell if in a future version further calls to _XReadEvents()  
> appear in the upstream Xlib code?

Please note, that libNX_X11 _XReadEvents() function will only ever  
used by nxagent. By nothing else. So we are free to handle things "as  
we want".

> So the real fix would be to find out why nx has added these return  
> to _XReadEvents() in the first place and if there are other ways to  
> get the desired effect without breaking the semantics.

What exactly do you mean by "breaking the semantics"? Can you point me  
to certain lines?

> Maybe instead of simply returning we could insert fake events.

Note that we are talking about events going through nxcomp. Events  
between client-side X.org Xserver <-> nxcomp <-> nxagent. Fake events  
sounds like extra bandwidth without meaningful payload.

## Mike

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22  0782 9AF4 6B30 2577 1B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

**Comment by uli42** [comment]
On  Do 23 Jun 2016 10:12:14 CEST, Ulrich Sibiller wrote:

> So would it make sense to introduce a looping around XPending()
> directly in nxagentWaitEvents()? Instead of leaving the blocking to
> NX_X11, we could move that into nxagent code directly.

I am not sure about adding a loop to nxagentWaitEvents. We'd need to anticipate what XReadEvents is going to do. So calling XPeekEvent (and therefore _XReadEvents) could be completely replaced by something own.

However, _XReadEvents is also called by MaskEvent, CheckIfEvent a.s.o. where special nx code has been added to handle the "emtpy" return. Maybe similar code has been forgotten in XPeekEvent.

> Please note, that libNX_X11 _XReadEvents() function will only ever
> used by nxagent. By nothing else. So we are free to handle things "as
> we want".

Ok, but do you have an overview what happens inside libNX_X11? I'd prefer to stay as close as possible to libX11 which will help when upgrading the code to newer Xlib versions.

> > So the real fix would be to find out why nx has added these return
> > to _XReadEvents() in the first place and if there are other ways to
> > get the desired effect without breaking the semantics.
> 
> What exactly do you mean by "breaking the semantics"? Can you point me
> to certain lines?

Yes. XlibInt.c contains this comment at the XReadEvents function header:

```
/* _XReadEvents - Flush the output queue,
 * then read as many events as possible (but at least 1) and enqueue them
 */
```

This definition is broken some lines later:

```
            /* find out how much data can be read */
            if (_X11TransBytesReadable(dpy->trans_conn, &pend) < 0)
#ifdef NX_TRANS_SOCKET
            {
                _XIOError(dpy);

                return;
            }
#else
                _XIOError(dpy);
#endif
```

Here a return statement has been added that is executed when no data (== no event) is there. So zero events have been read (and enqueued) which is contradicting  the function description.

Similar some line later:

```
#ifdef NX_TRANS_SOCKET
            if (dpy->flags & XlibDisplayIOError)
            {
#if defined(NX_TRANS_SOCKET) && defined(NX_TRANS_DEBUG)
                fprintf(stderr, "_XReadEvents: Returning with I/O error detected.\n");
#endif
                return;
            }
#endif
```

(this is the part that gets executed just before XPeekEvents segfaults).

> > Maybe instead of simply returning we could insert fake events.
> 
> Note that we are talking about events going through nxcomp. Events
> between client-side X.org Xserver <-> nxcomp <-> nxagent. Fake events
> sounds like extra bandwidth without meaningful payload.

We could simply add a fake event to dpy->head->events prior to returning from XReadEvents. That data would not travel the wire it would just satisfy XPeekEvent.

**Comment by sunweaver** [comment]
On  Do 23 Jun 2016 12:34:38 CEST, Ulrich Sibiller wrote:

> On  Do 23 Jun 2016 10:12:14 CEST, Ulrich Sibiller wrote:
> 
> > So would it make sense to introduce a looping around XPending()
> > directly in nxagentWaitEvents()? Instead of leaving the blocking to
> > NX_X11, we could move that into nxagent code directly.
> 
> I am not sure about adding a loop to nxagentWaitEvents. We'd need to  
> anticipate what XReadEvents is going to do. So calling XPeekEvent  
> (and therefore _XReadEvents) could be completely replaced by  
> something own.
> 
> However, _XReadEvents is also called by MaskEvent, CheckIfEvent  
> a.s.o. where special nx code has been added to handle the "emtpy"  
> return. Maybe similar code has been forgotten in XPeekEvent.

Ok, Then we should add "that code" there.

> > Please note, that libNX_X11 _XReadEvents() function will only ever
> > used by nxagent. By nothing else. So we are free to handle things "as
> > we want".
> 
> Ok, but do you have an overview what happens inside libNX_X11? I'd  
> prefer to stay as close as possible to libX11 which will help when  
> upgrading the code to newer Xlib versions.

Ohhhh... forget that. Latest libX11 is basically a wrapper around  
libxcb. It has changed so tremendously, you don't want to know.

> > > So the real fix would be to find out why nx has added these return
> > > to _XReadEvents() in the first place and if there are other ways to
> > > get the desired effect without breaking the semantics.
> > 
> > What exactly do you mean by "breaking the semantics"? Can you point me
> > to certain lines?
> 
> Yes. XlibInt.c contains this comment at the XReadEvents function header:
> 
> ```
> /* _XReadEvents - Flush the output queue,
>  * then read as many events as possible (but at least 1) and enqueue them
>  */
> ```

Maybe the above comment comes from the X.org team.

> This definition is broken some lines later:
> 
> ```
>             /* find out how much data can be read */
>             if (_X11TransBytesReadable(dpy->trans_conn, &pend) < 0)
> #ifdef NX_TRANS_SOCKET
>             {
>                 _XIOError(dpy);
> 
>                 return;
>             }
> #else
>                 _XIOError(dpy);
> #endif
> ```

And the code above here, of course, comes from NoMachine.

> Here a return statement has been added that is executed when no data  
> (== no event) is there. So zero events have been read (and enqueued)  
> which is contradicting  the function description.

I think, NoMachine forgot to clean up the comment. Not the first  
location where old X11 code is still in place.

> Similar some line later:
> 
> ```
> #ifdef NX_TRANS_SOCKET
>             if (dpy->flags & XlibDisplayIOError)
>             {
> #if defined(NX_TRANS_SOCKET) && defined(NX_TRANS_DEBUG)
>                 fprintf(stderr, "_XReadEvents: Returning with I/O  
> error detected.\n");
> #endif
>                 return;
>             }
> #endif
> ```
> 
> (this is the part that gets executed just before XPeekEvents segfaults).

Ok.

> > > Maybe instead of simply returning we could insert fake events.
> > 
> > Note that we are talking about events going through nxcomp. Events
> > between client-side X.org Xserver <-> nxcomp <-> nxagent. Fake events
> > sounds like extra bandwidth without meaningful payload.
> 
> We could simply add a fake event to dpy->head->events prior to  
> returning from XReadEvents. That data would not travel the wire it  
> would just satisfy XPeekEvent.

I'd rather immitate what is being done in MaskEvent, CheckIfEvent,  
etc. What has been done there to handle this, btw?

## Mike

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22  0782 9AF4 6B30 2577 1B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

**Comment by uli42** [comment]
> I'd rather immitate what is being done in MaskEvent, CheckIfEvent, etc. What has been done there to handle this, btw?

E.g. in PeekIfEvent.c and MaskEvent.c:

```
           _XReadEvents(dpy);
            if (prev && prev->qserial_num != qe_serial)
                /* another thread has snatched this event */
                prev = NULL;
#ifdef NX_TRANS_SOCKET
            if (_XGetIOError(dpy)) {
                return 0;
            }
#endif
```

at the end of the while loop. This is not applicable to XPeekEvent (and XNextEvent) since that is supposed to return an event and not just a number. I'll check this evening if leaving the dpy->head at NULL in case of an IOError and skipping the read is sufficient here.

Update: nxagentWaitEvents does not use the delivered event. It is just called to block (it seems). So it could be sufficient to replace XPeekEvent by some other event function.

**Comment by uli42** [comment]
Well, I am puzzled. It seems like I still have not fully understood the problem.

If have changed XPeekEvent like this:

```
diff --git a/nx-X11/lib/X11/PeekEvent.c b/nx-X11/lib/X11/PeekEvent.c
index b89c6c9..66e6871 100644
--- a/nx-X11/lib/X11/PeekEvent.c
+++ b/nx-X11/lib/X11/PeekEvent.c
@@ -46,7 +46,15 @@ XPeekEvent (dpy, event)
        LockDisplay(dpy);
        if (dpy->head == NULL)
            _XReadEvents(dpy);
+#ifdef NX_TRANS_SOCKET
+       if (_XGetIOError(dpy)) {
+         UnlockDisplay(dpy);
+         return 0;
+       }
+#else
        *event = (dpy->head)->event;
+#endif
+
        UnlockDisplay(dpy);
        return 1;
 }
```

And it is running rock solid now. However, the strange thing is: I still have debugging prints in _XReadEvents and while I see them when I do not add the patch above they completely vanish when I have the patch included.

In other words:

before: XReadEvents returns with IOError -> XPeekEvent crashes
after: XReadEvents does not execute the code with the IOError + the return

This is a bit unexpected. I have changed XPeekEvent to catch the error that happens in a called function and now the called function does not issue the error anymore. 

Also: while this code reliably prevents the crashes from happening it also issues "Xlib: sequence lost (0x10000 > 0xf7) in reply type 0x0!" right after a XSelect gets EINTR. The unpatched code segfaults instead...

Update: Looking at _XIOError: This was supposed to not return in original Xlib but the nx version is in fact returning. And that is probably the reason why XReadEvents adds a return statement after the _XIOError call.

Update2: Reexamining the segfault again I saw that the backtrace contained nxagentHandleConfigureNotify(). Which leads us back to the initial assumption that it has to do with the new Xinerama code. And - sadly - I can confirm this assumption was true. Here's what happens:
1. Unity's window manager repositions the nxagent windows everytime it is mapped. Why exactly it does that is out of scope here.
2. the original nx code is aware of WMs doing stuff like that. So it lets the WM do the positioning and only takes care of updating the position in the internal variables when it gets informed by a ConfigureNotifyEvent. The placement by the WM triggers such an event. 
3. the nxagent receives the ConfigureNotifyEvent for the nxagent window 
4. if the _size_ of the window changes the original nx code triggers randr adjustments. There's also an special event loop I have not fully understood yet (see below)
5. the newly added xinerama code is within the randr code.
6. since the new xinerama code also needs to recalculate stuff when the window position changes (window could cross monitor borders through movement) I simply extended the check mentioned in 4. to also trigger randr adjustments when the position changes. This change in Events.c is the reason for the segfault. I happens in line 3390 of the special event loop (nxagentWaitEvents is calling XPeekEvents):

```
   3381             NXFlushDisplay(nxagentDisplay, NXFlushLink);
   3382 
   3383             do
   3384             {
   3385               newEvents = False;
   3386 
   3387               timeout.tv_sec  = 0;
   3388               timeout.tv_usec = 500 * 1000;
   3389 
   3390               nxagentWaitEvents(nxagentDisplay, &timeout);
   3391 
   3392               /*
   3393                * This should also flush
   3394                * the NX link for us.
   3395                */
   3396 
   3397               XSync(nxagentDisplay, 0);
   3398 
   3399               while (XCheckTypedWindowEvent(nxagentDisplay, nxagentDefaultWindows[pScreen -> myNum],
   3400                                               ConfigureNotify, X))
   3401               {
   3402                 newEvents = True;
   3403               }
   3404 
   3405             } while (newEvents);
```

We can go back to the original code with this patch:

```
@@ -3370,9 +3370,7 @@ int nxagentHandleConfigureNotify(XEvent* X)
         if (nxagentOption(DesktopResize) == 1)
         {
           if (nxagentOption(Width) != X -> xconfigure.width ||
-                nxagentOption(Height) != X -> xconfigure.height ||
+                nxagentOption(Height) != X -> xconfigure.height )
-                 nxagentOption(X) != X -> xconfigure.x ||
-                nxagentOption(Y) != X -> xconfigure.y)
           {
             Bool newEvents = False;
```

Unfortunately this comes with a price:  xinerama is not working properly anymore. A simple window movement does not trigger a randr update now. I must check how I can achieve that again.

BTW: The newly introduced switch --rrxinerama will not help here...

**Comment by sunweaver** [comment]
On  Do 23 Jun 2016 23:39:47 CEST, Ulrich Sibiller wrote:

> Well, I am puzzled. It seems like I still have not fully understood  
> the problem.
> 
> If have changed XPeekEvent like this:
> 
> ```
> diff --git a/nx-X11/lib/X11/PeekEvent.c b/nx-X11/lib/X11/PeekEvent.c
> index b89c6c9..66e6871 100644
> --- a/nx-X11/lib/X11/PeekEvent.c
> +++ b/nx-X11/lib/X11/PeekEvent.c
> @@ -46,7 +46,15 @@ XPeekEvent (dpy, event)
>         LockDisplay(dpy);
>         if (dpy->head == NULL)
>             _XReadEvents(dpy);
> +#ifdef NX_TRANS_SOCKET
> +       if (_XGetIOError(dpy)) {
> +         UnlockDisplay(dpy);
> +         return 0;
> +       }
> +#else
>         *event = (dpy->head)->event;
> +#endif
> +
>         UnlockDisplay(dpy);
>         return 1;
>  }
> ```
> 
> And it is running rock solid now. However, the strange thing is: I  
> still have debugging prints in _XReadEvents and while I see them  
> when I do not add the patch above they completely vanish when I have  
> the patch included.
> 
> In other words:
> 
> before: XReadEvents returns with IOError -> XPeekEvent crashes
> after: XReadEvents does not execute the code with the IOError + the return
> 
> This is a bit unexpected. I have changed XPeekEvent to catch the  
> error that happens in a called function and now the called function  
> does not issue the error anymore.
> 
> Also: while this code reliably prevents the crashes from happening  
> it also issues "Xlib: sequence lost (0x10000 > 0xf7) in reply type  
> 0x0!" right after a XSelect gets EINTR. The unpatched code segfaults  
> instead...
> 
> Update: Looking at _XIOError: This was supposed to not return in  
> original Xlib but the nx version is in fact returning. And that is  
> probably the reason why XReadEvents adds a return statement after  
> the _XIOError call.
> 
> Update2: Reexamining the segfault again I saw that the backtrace  
> contained nxagentHandleConfigureNotify(). Which leads us back to the  
> initial assumption that it has to do with the new Xinerama code. And  
> - sadly - I can confirm this assumption was true. Here's what happens:
> 1. Unity's window manager repositions the nxagent windows everytime  
>    it is mapped. Why exactly it does that is out of scope here.
> 2. the original nx code is aware of WMs doing stuff like that. So it  
>    lets the WM do the positioning and only takes care of updating the  
>    position in the internal variables when it gets informed by a  
>    ConfigureNotifyEvent. The placement by the WM triggers such an event.
> 3. the nxagent receives the ConfigureNotifyEvent for the nxagent window
> 4. if the _size_ of the window changes the original nx code triggers  
>    randr adjustments. There's also an special event loop I have not  
>    fully understood yet (see below)
> 5. the newly added xinerama code is within the randr code.
> 6. since the new xinerama code also needs to recalculate stuff when  
>    the window position changes (window could cross monitor borders  
>    through movement) I simply extended the check mentioned in 4. to  
>    also trigger randr adjustments when the position changes. This  
>    change in Events.c is the reason for the segfault. I happens in line  
>    3390 of the special event loop (nxagentWaitEvents is calling  
>    XPeekEvents):
> 
> ```
>    3381             NXFlushDisplay(nxagentDisplay, NXFlushLink);
>    3382
>    3383             do
>    3384             {
>    3385               newEvents = False;
>    3386
>    3387               timeout.tv_sec  = 0;
>    3388               timeout.tv_usec = 500 * 1000;
>    3389
>    3390               nxagentWaitEvents(nxagentDisplay, &timeout);
>    3391
>    3392               /*
>    3393                * This should also flush
>    3394                * the NX link for us.
>    3395                */
>    3396
>    3397               XSync(nxagentDisplay, 0);
>    3398
>    3399               while (XCheckTypedWindowEvent(nxagentDisplay,  
> nxagentDefaultWindows[pScreen -> myNum],
>    3400                                               ConfigureNotify, X))
>    3401               {
>    3402                 newEvents = True;
>    3403               }
>    3404
>    3405             } while (newEvents);
> ```
> 
> We can go back to the original code with this patch:
> 
> ```
> @@ -3370,9 +3370,7 @@ int nxagentHandleConfigureNotify(XEvent* X)
>          if (nxagentOption(DesktopResize) == 1)
>          {
>            if (nxagentOption(Width) != X -> xconfigure.width ||
> -                nxagentOption(Height) != X -> xconfigure.height ||
> +                nxagentOption(Height) != X -> xconfigure.height )
> -                 nxagentOption(X) != X -> xconfigure.x ||
> -                nxagentOption(Y) != X -> xconfigure.y)
>            {
>              Bool newEvents = False;
> ```
> 
> Unfortunately this comes with a price:  xinerama is not working  
> properly anymore. A simple window movement does not trigger a randr  
> update now. I must check how I can achieve that again.
> 
> BTW: The newly introduced switch --rrxinerama will not help here...

Well done, Uli!!!

Mike

## 

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22  0782 9AF4 6B30 2577 1B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

**Comment by uli42** [comment]
I suppose that code would also crash if the WM would change the size instead of repositioning. So my additional check for position changes is not wrong, but the resulting code path is not SIGHUP safe because of XPeekEvent.

Therefore the fix in XPeekEvent is the way to go. There's no need to change xinerama then. I think the missing event is something we have to live with.

A similar patch could also be needed for XNextEvent. But I have found only one location where it is called and that is encapsulated by #ifdef DUMP, which is unset normally.

@HolySephi: Can you confirm the XPeekEvent patch is working for you?

**Comment by HolySephi** [comment]
``` diff
diff --git a/nx-X11/lib/X11/PeekEvent.c b/nx-X11/lib/X11/PeekEvent.c
index b89c6c9..66e6871 100644
--- a/nx-X11/lib/X11/PeekEvent.c
+++ b/nx-X11/lib/X11/PeekEvent.c
@@ -46,7 +46,15 @@ XPeekEvent (dpy, event)
        LockDisplay(dpy);
        if (dpy->head == NULL)
            _XReadEvents(dpy);
+#ifdef NX_TRANS_SOCKET
+       if (_XGetIOError(dpy)) {
+         UnlockDisplay(dpy);
+         return 0;
+       }
+#else
        *event = (dpy->head)->event;
+#endif
+
        UnlockDisplay(dpy);
        return 1;
 }
```

I'm unable to reproduce any segmentation faults with this patch too so i can confirm that this patch should resolve the segmentation fault.

[Edited by @Ionic: changed to diff syntax highlighting]

**Comment by sunweaver** [comment]
Hi Ulricht, Florian,

On  Fr 24 Jun 2016 10:06:53 CEST, Florian Wicke wrote:

> > diff --git a/nx-X11/lib/X11/PeekEvent.c b/nx-X11/lib/X11/PeekEvent.c
> > index b89c6c9..66e6871 100644
> > --- a/nx-X11/lib/X11/PeekEvent.c
> > +++ b/nx-X11/lib/X11/PeekEvent.c
> > @@ -46,7 +46,15 @@ XPeekEvent (dpy, event)
> >         LockDisplay(dpy);
> >         if (dpy->head == NULL)
> >             _XReadEvents(dpy);
> > +#ifdef NX_TRANS_SOCKET
> > -       if (_XGetIOError(dpy)) {
> > -         UnlockDisplay(dpy);
> > -         return 0;
> > -       }
> >   +#else
> >       *event = (dpy->head)->event;
> >   +#endif
> >   +
> >       UnlockDisplay(dpy);
> >       return 1;
> >   }
> 
> I'm unable to reproduce any segmentation faults with this patch too  
> so i can confirm that this patch should resolve the segmentation  
> fault.

@uli42: please file a PR for this fix. Thanks. And adapt XNextEvent,  
as well, even if rarely called.

## Mike

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22  0782 9AF4 6B30 2577 1B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de
